feat: Add document renderer order criteria events

aragon999 opened the pull request:
### 1. Why is this change necessary?
Currently it is not possible to modify the criteria objects, when creating documents, i.e. one needs to load the data afterwards in the next event.

### 2. What does this change do, exactly?
Add events to modify the criteria objects.

### 3. Describe each step to reproduce the issue or behaviour.
Extend the documents.

### 4. Please link to the relevant issues (if any).
\-

### 5. Checklist

- [x] I have rebased my changes to remove merge conflicts
- [ ] I have written tests and verified that they fail without my change
- [x] I have created a [changelog file](https://github.com/shopware/platform/blob/trunk/adr/2020-08-03-implement-new-changelog.md) with all necessary information about my changes
- [ ] I have written or adjusted the documentation according to my changes
- [ ] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfill them.

aragon999 committed:
feat: Add document renderer order criteria events

aragon999 commented on src/Core/Checkout/Document/Renderer/CreditNoteRenderer.php:
```diff
@@ -199,39 +199,48 @@ public function getDecorated(): AbstractDocumentRenderer
-        /** @var ?OrderEntity $order */
-        $order = $this->orderRepository->search($criteria, $versionContext)->get($orderId);
+        $criteria = OrderDocumentCriteriaFactory::create([$operation->getOrderId()], $rendererConfig->deepLinkCode, self::TYPE)
+            ->addFilter(new EqualsFilter('lineItems.type', LineItem::CREDIT_LINE_ITEM_TYPE));
```
Regarding the failing test, I would fix it like this
```suggestion

        $lineItemAssociation = $criteria->getAssociation('lineItems');
        $lineItemAssociation->addFilter(new EqualsFilter('type', LineItem::CREDIT_LINE_ITEM_TYPE));
```

However this would mean, that the order which is loaded with the version ID from the invoice, would be loaded, if a credit line item is present, and hence not the default version of the order would be loaded. But I would think this is the correct behavior.

Then in line 120 an error would be triggered, that no credit line items are found.

The current tests are working with that change, but as this would be an additional change in behavior, I thought I would mention it.

MalteJanz replied:
A change in behavior would be better mentioned in the changelog 😉 .

But I don't really get why some renders care about the order version id and others don't. Why does it matter for some? Is a new order version created when e.g. an invoice is generated, but the invoice renderer is using the live version or not? Maybe someone can enlighten me 🙈 @En0Ma1259 @mstegmeyer

aragon999 committed:
fixup: Only filter line items by type and not complete orders

En0Ma1259 commented on src/Core/Checkout/Document/Event/DocumentOrderCriteriaEvent.php:
```diff
@@ -0,0 +1,48 @@
+use Symfony\Contracts\EventDispatcher\Event;
+
+#[Package('checkout')]
+abstract class DocumentOrderCriteriaEvent extends Event
```
I think, we should get rid of these empty event classes with no benefit.
Make this class not abstract and add a name. Something like `Shopware\Core\Framework\DataAbstractionLayer\Event\EntityLoadedEvent`.
Use `Shopware\Core\Framework\Event\GenericEvent` interface
```php
    public function __construct(
        private readonly Criteria $criteria,
        private readonly Context $context,
        private readonly array $operations,
        private readonly DocumentRendererConfig $documentRendererConfig,
        string $rendererType
    ) {
        $this->name = $rendererType . '.document.criteria';
    }
````

aragon999 replied:
hey, sorry for the late reply, I was quite ill the last week, okay I will look into this, I thought it would be nicer, to have explicit classes to subscribe for different criteria events and I did it similar to the result events.

But I will have look how to optimize this.

aragon999 replied:
hey @En0Ma1259 sorry for the late reply and update of the PR, but now I had some time to look into it.

I implemented it a bit different, since what I would miss in that case, that I cannot use a class in a subscriber. Hence I used a constant on the `DocumentOrderCriteriaEvent`. 

The "drawback" is, that it is not made sure, that the event name has to end with `.document.criteria`. But I am not sure if a check in the constructor should be added to make sure, that this is the case, or if it is sufficient like that.

En0Ma1259 replied:
I don't think, we should do it that way. The event doesn't know where it will be dispatched and which document types exists.
For constant values, there are classes like `Shopware\Core\Content\Product\ProductEvents` with the event names. But will it really be necessary to create a class for this? We have other events where you need to set a string key inside a subscriber.

And as you mentioned, we can't be sure, if the event ends with `.document.criteria`. That should be the most important point. A separate check would only add more logic, where it isn't needed.


```php
    public function __construct(
        private readonly Criteria $criteria,
        private readonly Context $context,
        private readonly array $operations,
        private readonly DocumentRendererConfig $documentRendererConfig,
        private readonly string $documentType,
    ) {
        $this->name = $documentType . '.document.criteria';
    }
```
And should be called then with 
```php
$this->eventDispatcher->dispatch(new DocumentOrderCriteriaEvent(
    $criteria,
    $context,
    $operations,
    $rendererConfig,
    self::TYPE
));
```

aragon999 replied:
I changed it as suggested and left the constants in the event, what do you think about this solution?

En0Ma1259 commented on src/Core/Checkout/Document/Event/DocumentOrderCriteriaEvent.php:
```diff
@@ -0,0 +1,60 @@
+use Shopware\Core\Framework\Log\Package;
+use Symfony\Contracts\EventDispatcher\Event;
+
+#[Package('checkout')]
```
We have changed the package internally. It is now `after-sales`

aragon999 replied:
I adjusted the `Package`

En0Ma1259 commented on src/Core/Checkout/Document/Event/DocumentOrderCriteriaEvent.php:
```diff
@@ -0,0 +1,60 @@
+    public const STORNO_ORDER_CRITERIA_EVENT = 'storno.document.criteria';
+
+    /**
+     * @param DocumentGenerateOperation[] $operations
```
With our latest change, this is now
```suggestion
     * @param array<string, DocumentGenerateOperation> $operations
```

Would you be so kind and change the annotation [here](https://github.com/shopware/shopware/blob/trunk/src/Core/Checkout/Document/Event/DocumentOrderEvent.php#L25) as well. This place has been forgotten

En0Ma1259 commented on src/Core/Checkout/Document/Event/DocumentOrderCriteriaEvent.php:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    /**
+     * @return DocumentGenerateOperation[]
```
```suggestion
     * @return array<string, DocumentGenerateOperation>
```

MalteJanz commented:
Please update from `trunk` and resolve all threads

aragon999 commented:
I just rebased against trunk and implemented the comments. As this PR now contains also a bug fix (i.e. the filtering of the line items of the credit note), I am not sure if this should be split into two PRs?

MalteJanz commented:
@aragon999 I'm fine with leaving the small bug fix in here (just add it to the changelog). But if you really want you can also create another PR for just that.

Overall the PR already looks quite good 👍 , I would only like to see some tests, e.g.:
- for the bug fix
- that the new event is properly dispatched in all renderers and properly affects the criteria used in the order repository by the renderers

aragon999 commented:
@MalteJanz no I do not want to create another PR just for the sake of another PR :-D It was more like if it helps to have it in two different commits.

I will look into the tests, but the current tests are actually failing with the "bugfix" as the current implementation works around this, by just using no filter of the line item type in the second criteria: https://github.com/shopware/shopware/blob/76bc50f7f3e0fd23b7a35ecc3cb155346a3a8708/src/Core/Checkout/Document/Renderer/CreditNoteRenderer.php#L241-L244

I will check if I can improve the tests there.

En0Ma1259 commented on src/Core/Checkout/Document/Event/DocumentOrderCriteriaEvent.php:
```diff
@@ -0,0 +1,67 @@
+#[Package('after-sales')]
+final class DocumentOrderCriteriaEvent extends Event implements GenericEvent
+{
+    public const CREDIT_NOTE_ORDER_CRITERIA_EVENT = CreditNoteRenderer::TYPE . '.document.criteria';
+    public const DELIVERY_ORDER_CRITERIA_EVENT = DeliveryNoteRenderer::TYPE . '.document.criteria';
+    public const INVOICE_ORDER_CRITERIA_EVENT = InvoiceRenderer::TYPE . '.document.criteria';
+    public const STORNO_ORDER_CRITERIA_EVENT = StornoRenderer::TYPE . '.document.criteria';
```
Remove these const values pls. They are not used anymore

aragon999 replied:
I explicitly adjusted them, to have the actual events, that they can be used like the `ProductEvents` constants, i.e. that they can be used in a plugin to subscribe to an event.

En0Ma1259 replied:
A plugin can also use a normal string. Not all event names have a constant value, and there is no event class with the event names as const values.
As I commented before 
> For constant values, there are classes like `Shopware\Core\Content\Product\ProductEvents` with the event names.

If you want these const values anywhere, create a new class `Shopware\Core\Checkout\Document\DocumentEvents`

MalteJanz replied:
@aragon999 can you adjust to this please 🙂

aragon999 replied:
Ah yes, sorry I missed the latest comment. I will do that today :-)

aragon999 replied:
I moved the constants

aragon999 committed:
fixup: Move event constants to own class

aragon999 committed:
fixup: Codestyle and phpstan

aragon999 committed:
fixup: Do not assert that newly generated order has the live version
